traffic_mirroring.go

```diff
@@ -30,36 +26,85 @@ func main() {
 	}
 
 	packetSource := gopacket.NewPacketSource(handle, handle.LinkType())
-	max_size := 10
-	packets_list := make([][]byte,max_size) 
-	current_packet := 0
+	packetsList := make(chan []byte, 22)
+	filePath := "dump_file"
+	batchSize := 20
+
+	f, err := os.Create(filePath)
+	check(err)
+	defer f.Close()
+
+	sigChan := make(chan os.Signal)
+	quit := make(chan bool)
+	signal.Notify(sigChan, os.Interrupt)
+	go handleSignal(sigChan, packetsList, quit, os.Exit)
+	go WriteFile(packetsList, f, quit, batchSize)
 
 	for overlayPacket := range packetSource.Packets() {
 		vxlanLayer := overlayPacket.Layer(layers.LayerTypeVXLAN)
-		if vxlanLayer != nil {
-			vxlanPacket, _ := vxlanLayer.(*layers.VXLAN)
-			
-			fmt.Println(current_packet)
-			packets_list[current_packet] = vxlanPacket.LayerPayload()
-			current_packet += 1
-			fmt.Println(vxlanPacket.LayerPayload())
-			if current_packet == max_size {
-				newSlice := make([][]byte, max_size)
-				copy(newSlice, packets_list)
-				fmt.Println(current_packet)
-				current_packet = 0
-				go write_file(newSlice, f)
+		if vxlanLayer == nil {
+			log.Printf("Unable to get VXLAN Layer for packet with metadata (%+v)\n", overlayPacket.Metadata())
+		}
+		vxlanPacket, ok := vxlanLayer.(*layers.VXLAN)
+		if !ok {
+			log.Printf("Unable to cast packet (%+v) to vxlan layer", overlayPacket.Metadata())
+		}
+		packetsList <- vxlanPacket.LayerPayload()
+	}
 
+}
+
+
+func check(e error) {
+	if e != nil {
+		panic(e)
+	}
+}
+
+func handleSignal(sigChan chan os.Signal, packetsList chan []byte, quit chan bool, end func(int)) {
+	<-sigChan
+	exitCode := 0
+	close(packetsList)
+	quit <- true
+	end(exitCode)
+}
 
+func NetworkListener(source gopacket.PacketSource, dest chan []byte) {
+	for overlayPacket := range source.Packets() {
+		vxlanLayer := overlayPacket.Layer(layers.LayerTypeVXLAN)
+		if vxlanLayer != nil {
+			vxlanPacket, ok := vxlanLayer.(*layers.VXLAN)
+			if !ok {
+				log.Printf("Unable to cast packet to vxlan layer")
+			} else {
+				dest <- vxlanPacket.LayerPayload()
 			}
 		}
 	}
-	defer f.Close()
 }
 
-func write_file(packets_list [][]byte, f *os.File) {
-	for _, value := range packets_list {
-		fmt.Fprintln(f, value)  
+func WriteFile(packetsList chan []byte, file io.Writer, quit chan bool, batchSize int) {
+
+	for {
+		select {
+		case <- quit:
+			for value := range packetsList {
+				_, err := file.Write(value)
+				if err != nil {
+					panic(err)
+				}
+			}
```

Comment:
After writing the remaining packets, you need to stop this method execution, you can do it by returning right here or break the `for` infinite loop.